web/pdf_history.js

```diff
@@ -13,21 +13,455 @@
+  }
+
+  /**
+   * @private
+   */
+  _isValidState(state) {
+    if (!state) {
+      return false;
+    }
+    if (state.fingerprint !== this.fingerprint) {
+      // This should only occur in viewers with support for opening more than
+      // one PDF document, e.g. the GENERIC viewer.
+      return false;
+    }
+    if (!Number.isInteger(state.uid) || state.uid < 0) {
+      return false;
+    }
+    if (typeof state.destination !== 'object' || state.destination === null) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @private
+   */
+  _resetSomeInternalState(destination, uid) {
+    this._destination = destination;
+    this._currentUid = uid;
+    this._uid = this._currentUid + 1;
+  }
+
+  /**
+   * @private
+   */
+  _updateViewarea({ location, }) {
+    this._position = {
+      hash: this._isViewerInPresentationMode ?
+        `page=${location.pageNumber}` : location.pdfOpenParams.substring(1),
+      page: this.linkService.page,
+      first: location.pageNumber,
+    };
+
+    if (this._popStateInProgress) {
+      return;
+    }
   }
 
-  initialize(fingerprint, resetHistory = false) {}
+  /**
+   * @private
+   */
+  _popState({ state, }) {
+    let newHash = getCurrentHash(), hashChanged = this._currentHash !== newHash;
+    this._currentHash = newHash;
+
+    if (!state ||
+        (typeof PDFJSDev !== 'undefined' && PDFJSDev.test('CHROME') &&
+         state.chromecomState && !this._isValidState(state))) {
+      // This case corresponds to the user changing the hash of the document.
+      this._currentUid = this._uid;
 
-  push() {}
+      let { hash, page, } = parseCurrentHash(this.linkService);
+      this._pushOrReplaceState({ hash, page, }, /* forceReplace */ true);
+      return;
+    }
+    if (!this._isValidState(state)) {
+      // This should only occur in viewers with support for opening more than
+      // one PDF document, e.g. the GENERIC viewer.
+      return;
+    }
+
+    // Prevent the browser history from updating until the new destination,
+    // as stored in the browser history, has been scrolled into view.
+    this._popStateInProgress = true;
+
+    if (hashChanged) {
+      // When the hash changed, implying that the 'popstate' event will be
+      // followed by a 'hashchange' event, then we do *not* want to update the
+      // browser history when handling the 'hashchange' event (in web/app.js)
+      // since that would *overwrite* the new destination navigated to below.
+      //
+      // To avoid accidentally disabling all future user-initiated hash changes,
+      // if there's e.g. another 'hashchange' listeners that stops the event
+      // propagation, we make sure to always force-reset `this._blockHashChange`
+      // after `HASH_CHANGE_TIMEOUT` milliseconds has passed.
+      this._blockHashChange++;
+      waitOnEventOrTimeout({
+        target: window,
+        name: 'hashchange',
+        delay: HASH_CHANGE_TIMEOUT,
+      }).then((type) => {
```

Comment:
This is now an unused variable. Usually this is denoted by replacing `type` with `_` (indicating we don't care about what it is).

Reply:
I'll just remove it completely, i.e. `}).then(() => {` instead.